### ISSUE TITLE
Toolchains

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#7432]
*This proposal is an alternative to #7199. The problem is the same, but the proposed solution is different.*

## Problem

There are two types of Dagger modules: those that are standalone software projects, and those that exist in the context of another software project. Let's call those *standalone modules* and *contextual modules*, respectively.

Dagger is designed to support both, which is good, but causes friction for users in some areas.

- Contextual modules almost always need access to their context directory. Today this requires an explicit argument, (which usually looks like `dagger call MYFUNC --source=.`), which is verbose.
- Contextual modules often need to filter their context directory, usually to avoid the prohibitive cost of uploading a huge monorepo on every task.
- The best practice for contextual modules is to make the context directory itself a module, and store the module source code in a configurable subdirectory. This additional concept of "module source" being different from "module root", adds complexity for the user.
- For standalone modules, the recommended value for `source` is `.`. For contextual modules, it is `./dagger`. The current default value is `./dagger`, which hurts the experience of creating a standalone module. If we change it to `.`, it will hurt the experience of creating contextual modules. Either way, the user experience suffers.

## Solution

I propose introducing the concept of *toolchain* as a first-class concept in Dagger.

## What is a toolchain?

According to gpt4o:

> A toolchain is a set of programming tools used to perform complex software development tasks, typically including:
>
> 1. Compiler: Translates source code into executable code.
> 2. Linker: Combines object files into a single executable or library.
> 3. Debugger: Helps identify and fix bugs.
> 4. Version Control: Manages changes to source code over time.
> 5. Build Automation: Automates the build process, often using tools like Make or CI/CD systems.
>
> Each tool in the chain works together to streamline development and deployment.

Every software project has a toolchain. And every daggerized project has a toolchain implemented as Dagger module.

## Daggerizing a toolchain

When a project adopts Dagger, it designates a Dagger module as that project's toolchain. The toolchain could be purpose-built for the project (eg the`./ci` module in `github.com/dagger/dagger`) or more generic (for example, a Docker toolchain that understands `Dockerfile` and `docker-compose.yaml`; or a Ruby-on-Rails toolchain for developing RoR apps). Most projects start using an off-the-shelf toolchain, then eventually create their own as the project becomes more complex. Dagger can support both equally well.

### How it works

A Dagger module can designate another module as its toolchain. This is done with the `toolchain` field in `dagger.json`. Any module can be used as a toolchain for another module.

The rules are for the same as for dependencies:

- Local modules are allowed, but restricted to the same git repository
- Remote modules must be pinned

Example of a project embedding its own toolchain:

```json
{
 "name": "dagger",
 "toolchain": "./ci"
}
```

Example of a project using a remote toolchain:

```json
{
 "name": "myapp",
 "toolchain": "github.com/shykes/docker@51b4f2a62d6a1c71cf5ab3def907ed262949b3b0"
}
```

### Selecting a toolchain

A toolchain can be selected when initializing the module:

```console
dagger init --toolchain=./ci
```

```console
dagger init --toolchain=github.com/shykes/docker
```

### Toolchain vs SDK

A module can have a toolchain or an SDK, but not both.

- `--sdk` and `--toolchain` are mutually exclusive flags to `dagger init`
- `sdk` and `toolchain` are mutually exclusive fields in `dagger.json`

This is because SDK and Toolchain are 2 different ways of populating the module with functions:

- Using an SDK means: "this module implements its own Dagger functions. Use this SDK to load them"
- Using a toolchain means: "this module relies on another module to implement its Dagger functions. Use this toolchain to load *them*"

For example, in the case of Dagger:

- `github.com/dagger/dagger` uses a toolchain: `./ci`
- `github.com/dagger/dagger/ci` uses a SDK (Go?)

### Access to the context directory

This part is the same as #7199 . A toolchain module can access its caller's context via `dag.Context.Directory()`.

**Comment by vito** [comment]
Let's say I bootstrap my Rails app and use the off-the-shelf Rails toolchain. Next week I want to add my own custom function. What does that look like?

I ask because it seems inevitable for most real-world projects, and if we don't have a good story the "reusable toolchain" aspect will feel more like a parlor trick. For example, if I have to start over with an SDK and manually re-add all the Rails functions as pass-throughs to the Rails toolchain, that's no fun. And even if it's not inevitable, it doesn't seem like we'd want users to learn that it's hard to add functions to their project if they use a toolchain.

Personally I have some skepticism that the Rails case really fits the same bucket as the `./ci` use case. To me it just feels like a plain old module that takes your code as an input, whereas the `./ci` use case is more tightly intertwined.

It would also be bad IMO if we end up with two competing "flavors" of modules (toolchain vs. vanilla) that have to be invoked in a different way (context dir vs. dir arg). Module consumers will have to know what type of module it is, and module authors will have to decide whether it's exposed as a toolchain or not, which seems non-obvious as it'll come up any time there's a `Directory` involved, which is often.

(Sorry for the mostly wet blanket comments! Just poking holes to keep things moving. Maybe there are clean answers to this that aren't outlined yet.)

**Comment by shykes** [comment]
> Let's say I bootstrap my Rails app and use the off-the-shelf Rails toolchain. Next week I want to add my own custom function. What does that look like?

I think either fork, wrap, or reimplement it.

- If you need to customize the off-the-shelf toolchain beyond what it can be configured to do: **fork**
- If you need to abstract it away, for example you now have a RoR backend + a NextJS frontend, and also want to manage the CDN: wrap into a higher-level toolchain that provides more abstraction and/or more breadth
- Beyond a certain level of customization, you would just reimplement it, possibly reusing the lower-level modules (ie. maybe the RoR module imported a great Ruby module, and you want to call that directly)

> Personally I have some skepticism that the Rails case really fits the same bucket as the `./ci` use case. To me it just feels like a plain old module that takes your code as an input, whereas the `./ci` use case is more tightly intertwined. (For example, it's versioned as part of your repo, as opposed to a Rails toolchain module which is versioned independently.)

I think the "Rails" example is realistic *but* it would need to be opinionated enough to, say, include deployment options out of the box.

The question of how "intertwined" with the application is key. In my experience, that question is *always* an issue as soon as you want some sort of reusable "platform" across projects. There's always the problem of *drawing the line*. A typical example: can the app provide its own Dockerfile? Or does the platform lock one in?

I think with the concept of toolchain module + access to the context directory, it becomes possible to define the "line" you want as code.

> It would also be bad IMO if we end up with two competing "flavors" of modules (toolchain vs. vanilla) that have to be invoked in a different way (context dir vs. dir arg) but are otherwise 99% the same. Module consumers will have to know what type of module it is, and module authors will have to decide whether it's exposed as a toolchain or not, which seems non-obvious as it'll come up any time there's a `Directory` involved, which is often.

Yes I agree that is a concern. Note that it's exactly the same concern in the [Contextual modules proposal](#7199).

Possible answers:

1. It's not that bad, in practice it works
2. You can label your module as `is_toolchain` which basically is a non-binding promise that your module might call `dag.Context().Directory()`
3. We change context directory API, so that it can be discovered statically. For example it could be a marker on a `Directory` arguments. If your directory argument has `//+context`, it defaults to the current context - and your module is marked as a toolchain.

**Comment by sipsma** [comment]
> This is because SDK and Toolchain are 2 different ways of populating the module with functions:
> * Using an SDK means: "this module implements its own Dagger functions. Use this SDK to load them"
> * Using a toolchain means: "this module relies on another module to implement its Dagger functions. Use this toolchain to load them"

I don't think this alone really distinguishes a toolchain from an SDK. An SDK today is a module that takes an arbitrary directory and turns it into a set of functions. The existing SDKs happen to do that by parsing code that lives in those directories, but you could already today implement an SDK that parses anything arbitrary in a directory (Dockerfile, compose.yaml, rails app, etc. etc.) and turn it into a set of functions.

Given that, it seems like the new thing that a toolchain does is just allow access to the context directory. But at that point it seems like all we need is to make whether a given module has access to the context directory something that's configurable; no new high level concepts required. What's missing if we just do that?

I'll echo Alex's comment that I'm not trying to be a wet blanket here 🙂 Just trying to see if we can whittle down a design here to the true bare essentials, but will be happy to realize I'm missing something essential.

**Comment by shykes** [comment]
Yes the thought occurred to me that SDKs and toolchains are awfully similar... But couldn't find a way to merge the concepts without making things confusing in practice. Would be happy to try again though!

Also the word "toolchain" (credit to @vito) really struck me as a great fit.

**Comment by vito** [comment]
> * If you need to customize the off-the-shelf toolchain beyond what it can be configured to do: **fork**
> 
>     * If you need to abstract it away, for example you now have a RoR backend + a NextJS frontend, and also want to manage the CDN: wrap into a higher-level toolchain that provides more abstraction and/or more breadth
> 
>     * Beyond a certain level of customization, you would just reimplement it, possibly reusing the lower-level modules (ie. maybe the RoR module imported a great Ruby module, and you want to call that directly)

* This sounds sensible if your new function fits the toolchain, and it encourages contributing a PR upstream for others' benefit, so :+1:
* This is an interesting case - can my module use _mutliple_ toolchains? What does that look like? Seems like it wouldn't be uncommon. My guess is you'd define an entrypoint for each toolchain, and you could have multiple instances of them exposed as functions on your own module/toolchain. But how do you do that? Modules can't return types from other modules, so I don't think you can stitch together hierarchies like that at the moment.
* This is the step I was afraid of; reimplementing the toolchain sounds like a lot of work, and has the downside of no longer benefitting from upstream toolchain updates. Ideally I'd still be able to start with a toolchain and then add a random function without feeling like I'm in a toolchain prison.

Feels like there's a mixin/layering system hiding in here somewhere. :thinking:

**Comment by sipsma** [comment]
> This is an interesting case - can my module use mutliple toolchains? What does that look like? Seems like it wouldn't be uncommon. 

I think that could be modeled as enabling support for multiple SDKs in `dagger.json`, basically so it's a list rather than just a single SDK.

But I guess I'm missing how this is connected to the problem at hand, though I enjoy the idea nonetheless.

Also bumping the crux of my previous comment:
> Given that, it seems like the new thing that a toolchain does is just allow access to the context directory. But at that point it seems like all we need is to make whether a given module has access to the context directory something that's configurable; no new high level concepts required. What's missing if we just do that?

**Comment by shykes** [comment]
@vito toolchains are just modules. Basically this is a convenience to avoid having to mess around with `-m ./ci`, `source` configuration etc.

So mixin/layering of toolchains just means mixin/layering of modules. Sounds like sticking to tried-and-true software development workflows is best (fork; wrap; reimplement; all these things apply equally to a dagger module or a go library)

**Comment by shykes** [comment]
> But at that point it seems like all we need is to make whether a given module has access to the context directory something that's configurable; no new high level concepts required. What's missing if we just do that?

I think we would still be missing a way to tell dagger *which* module should be run in the context of *which* other module, and in a way that is convenient for end-users.

eg. `git clone foo; cd foo; dagger functions`

**Comment by vito** [comment]
(Let me know if I'm missing something obvious, engaging in this discussion pretty late :sweat_smile:)

> I think that could be modeled as enabling support for multiple SDKs in `dagger.json`, basically so it's a list rather than just a single SDK.
> 
> But I guess I'm missing how this is connected to the problem at hand, though I enjoy the idea nonetheless.

The connection is that if someone uses a Daggerverse module as their toolchain (e.g. Rails), and you can only have one toolchain, the user is locked in to that toolchain and can't just add their own functions - which is something that's really easy today and part of the charm of the DX. The steps described in https://github.com/dagger/dagger/issues/7432#issuecomment-2123436127 are essentially to a) fork the toolchain, b) compose it, or c) scrap it, only b) doesn't seem possible right now.

If it were an array with mixin/layering-like semantics you could just write your own module `./foo` and add to the array, which is nice, but if you were to use multiple toolchains they'd probably compete for `build`/`test`/etc., so that doesn't seem like a good fit.

Defining your own functions as entrypoints to toolchain modules seems more robust, but that gets back to the 'modules can't return types from other modules' issue.

> Also bumping the crux of my previous comment:
> 
> > Given that, it seems like the new thing that a toolchain does is just allow access to the context directory. But at that point it seems like all we need is to make whether a given module has access to the context directory something that's configurable; no new high level concepts required. What's missing if we just do that?

I think there are two "new" parts to it:

1. Being able to bootstrap a project with a suite `dagger call` commands.
2. Automatically passing my project's context to modules that are installed in a certain way.

Arguably 1) is already possible by installing a module and invoking it from its local name (`dagger -m rails build`). I say arguably because another implicit goal here is passing fewer flags around.

And as you said 2) seems to be the main difference, but I think we'd need to be careful here; a module probably shouldn't be able to opt itself into access to the context dir. Ideally it gains access based on how it's installed or used.

But it's also 2) that I'm pretty worried about (the "two different module flavors" point before).

Dumb idea: can't we avoid the whole "contextual" vs "non-contextual" modules issue by just making the context dir an automatically-supplied arg from the CLI? That way it's purely a convention; there are no more module flavors, just CLI sugar, which is where the pain is felt. If the Rails module took a `context` arg, `dagger -m rails build` would "just work", and then one day we can figure out how to compose modules together.

**Comment by sipsma** [comment]
> And as you said 2) seems to be the main difference, but I think we'd need to be careful here; a module probably shouldn't be able to opt itself into access to the context dir. Ideally it gains access based on how it's installed or used.

Why is that? My understanding is that the whole goal here is to make daggerizing a project easier. If you want to daggerize your project, then you want your module's functions to have access to the context of that project. So what I'm suggesting is that when initializing it you just enable access to the context dir (`dagger init --context ...`). What problems are introduced with that approach?

> Being able to bootstrap a project with a suite dagger call commands.

I do also want us to have ways of doing fancier "templated" inits so you can start out with something more than just the silly "hello world"-esque starter functions you get today.

I'll point out that this would already be possible to implement as a module today by writing a function that accepts a `Directory` as an argument and returns a `Directory` that you can just export
* i.e. starting from an un-daggerized project run `dagger call -m github.com/dagger/inits/rails init --source . export --path .`

But I also nonetheless see how this could get integrated into the CLI commands in a cleaner way and/or be integrated with SDKs. So I need to do some more mulling on this whole point 🤔

---

> I think we would still be missing a way to tell dagger which module should be run in the context of which other module, and in a way that is convenient for end-users.

But why would users want to run their module in the context of another module vs. just optionally have access to the context of their project that they are daggerizing?

My overarching concern here is I just don't want to add too much conceptual overhead to everything. Personally, whenever I start trying to use some tool and need to first review a huge glossary of concepts and terminology my eyes tend to glaze over and I just don't use it.

There's a balance to be found here of course, you can't just have no concepts, but I'm trying to see if we can keep the "conceptual creep" to a minimum. If not, then we'll go from there, but that's where I'm coming from.
* e.g. The conceptual overhead of "daggerizing a project requires initializing a module that's configured to have access to its context directory" seems much smaller to me than "daggerizing a project requires initializing a module that uses a toolchain which itself is another module and where your functions are actually implemented"

**Comment by shykes** [comment]
@TomChv it's the same mechanism as in #7199 . Just named differently and a different layout in the repo.

**Comment by shykes** [comment]
To be clear "toolchain module" is an attempt at making "entrypoint module" more appealing and intuitive. If the word "entrypoint" makes it feel less difficult to understand than "toolchain", that's fine with me.

Other words I considered, beyond "entrypoint" and "toolchain":

- SDK
- Framework
- Platform

**Comment by vito** [comment]
> To be clear "toolchain module" is an attempt at making "entrypoint module" more appealing and intuitive. If the word "entrypoint" makes it feel less difficult to understand than "toolchain", that's fine with me.

And to be fair, in our side discussion I initally had a negative reaction to "entrypoint" since I was afraid of over-using it. :sweat_smile: Sorry for the bait and switch. Since we're already using that term for "module entrypoints" and it's [clearly called out in Daggerverse](https://daggerverse.dev/mod/github.com/purpleclay/daggerverse/golang@1190a20046659b68001a6e16cd5c12fab960b216#golang.constructor), it seems worth continuing to use if the proposal is really building on the exact same thing.

**Comment by sipsma** [comment]
> I'm realizing we seem to all have competing proposals in our heads and we're only showing bits and pieces of them which leads to talking past each other, so I'll try and paint a full picture of how I'm thinking about this. (Maybe I'm missing some context from the original discussion - sorry)

Good call, I also feel like I'm missing context from another discussion 😄 We're gonna discuss this live but will write this out ahead of time to help concretize.

My original thinking:

Daggerizing a project only requires authoring a module, same as authoring a standalone module, the only difference is that you initialize the module to have access to the context dir of the project you are daggerizing.

So you'd run e.g. `dagger init --context=true .` in the root of your repo (or `dagger init --context=true --source=ci/`, etc. depending on where you want the module source code to live)
* Sorry about the confusion in my previous comment, I meant `...` as in ellipsis for the rest of the command, but in retrospect should have just typed it 😅

If you do that, then `dag.Context().Directory()` returns the context dir. If you don't then `dag.Context().Directory()` returns an empty dir.
* Basically the same sort of idea as https://github.com/dagger/dagger/issues/7199 IIUC
* You'd also need some way of configuring include/exclude on the context dir still, which would be configured in `dagger.json` (unfortunate, but just technical reality atm)

That's it. No need for a separate module or separate concepts (whatever the terminology). Writing a module that's daggerizing a project is the same as writing any standalone module except you have access to the context of the project you are daggerizing.

---

That's the original thought, but based on the discussion here:
1. I'd not be opposed to what you mentioned @vito of using a special entrypoint argument instead of configuration at `dagger init`-time.
   * In my mind, that's trading one piece of magic (whether `dag.Context().Directory()` returns something) for another piece of magic (whether the entrypoint has that special arg), so it's mostly a wash IMO.
   * If we do go with the arg approach, things worth considering:
      * Making it magic in the *type* of the arg rather than the name of the argument. So a special `ContextDirectory` type or something. That just feels a bit more resilient in terms of handling consistently across all SDKs.
      * Consider allowing the include/exclude to be set on the arg rather than in `dagger.json` (i.e. a pragma, `@annotation`, etc.). Just so we centralize this in one place rather than split it between `dagger.json` and the entrypoint func.
2. I agree with the value of some ability to inherit/mixin/etc. functions rather than always needing to explicitly write them yourself
   * However, at the moment I have a strong sense that this should be an orthogonal feature from just improving the experience of daggerizing a project and having access to the context dir. The main reason being that it seems like it would be useful for standalone modules too.
   * But I'm not super sure on this yet, very open to it somehow being related enough to be worth intertwining here. Need to think+discuss more.

**Comment by shykes** [comment]
Update: we had some out-of-band discussions with @sipsma and @vito, which I am still digesting.

**Comment by rawkode** [comment]
It's been a few weeks, are we any further forward with this or #7199?

I'm in a pickle with service local modules calling shared modules and needing a way to handle passing the context down and these proposals appear to be trying to solve that problem.

**Comment by shykes** [comment]
> It's been a few weeks, are we any further forward with this or #7199?
> 
> I'm in a pickle with service local modules calling shared modules and needing a way to handle passing the context down and these proposals appear to be trying to solve that problem.

We have consensus on a subset, which I just wrote down here: https://github.com/dagger/dagger/issues/7647

**Comment by shykes** [comment]
I am closing this in favor of the more narrowly scoped [context directory proposal (#7647)](#7647). Once that is shipped, we can revisit this issue, and look for the next chunk to ship.
